File: src/server/services/marketData.js
import { EventEmitter } from 'events';

// Market assets configuration
const ASSETS = {
  // Equities
  SPY: { name: 'S&P 500 ETF', type: 'equity', basePrice: 592.50, volatility: 0.008 },
  QQQ: { name: 'Nasdaq 100 ETF', type: 'equity', basePrice: 518.30, volatility: 0.012 },
  IWM: { name: 'Russell 2000 ETF', type: 'equity', basePrice: 225.40, volatility: 0.015 },
  DIA: { name: 'Dow Jones ETF', type: 'equity', basePrice: 432.80, volatility: 0.007 },

  // FX
  'EUR/USD': { name: 'Euro/Dollar', type: 'fx', basePrice: 1.0298, volatility: 0.002, decimals: 4 },
  'GBP/USD': { name: 'Pound/Dollar', type: 'fx', basePrice: 1.2198, volatility: 0.003, decimals: 4 },
  'USD/JPY': { name: 'Dollar/Yen', type: 'fx', basePrice: 157.85, volatility: 0.003, decimals: 2 },
  DXY: { name: 'Dollar Index', type: 'fx', basePrice: 109.65, volatility: 0.002 },

  // Bonds
  TLT: { name: '20+ Year Treasury ETF', type: 'bond', basePrice: 87.45, volatility: 0.008 },
  IEF: { name: '7-10 Year Treasury ETF', type: 'bond', basePrice: 93.20, volatility: 0.004 },
  HYG: { name: 'High Yield Corp Bond', type: 'bond', basePrice: 77.85, volatility: 0.005 },

  // Volatility
  VIX: { name: 'CBOE Volatility Index', type: 'volatility', basePrice: 16.80, volatility: 0.05 },
  VVIX: { name: 'VIX of VIX', type: 'volatility', basePrice: 92.50, volatility: 0.08 },

  // Commodities
  GLD: { name: 'Gold ETF', type: 'commodity', basePrice: 242.30, volatility: 0.006 },
  USO: { name: 'Oil Fund', type: 'commodity', basePrice: 73.45, volatility: 0.02 },
};

export class MarketDataService extends EventEmitter {
  constructor() {
    super();
    this.assets = ASSETS;
    this.currentPrices = {};
    this.priceHistory = {};
    this.intervalId = null;

    // Initialize prices
    Object.entries(ASSETS).forEach(([symbol, config]) => {
      this.currentPrices[symbol] = {
        symbol,
        name: config.name,
        type: config.type,
        price: config.basePrice,
        open: config.basePrice,
        high: config.basePrice,
        low: config.basePrice,
        change: 0,
        changePercent: 0,
        volume: Math.floor(Math.random() * 1000000),
        lastUpdate: new Date().toISOString()
      };
      this.priceHistory[symbol] = [{ time: Date.now(), price: config.basePrice }];
    });
  }

  startSimulation() {
    // Update prices every 500ms
    this.intervalId = setInterval(() => {
      this.updatePrices();
    }, 500);

    console.log('Market data simulation started');
  }

  stopSimulation() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
  }

  updatePrices() {
    const updates = [];

    Object.entries(this.assets).forEach(([symbol, config]) => {
      const current = this.currentPrices[symbol];

      // Random walk with mean reversion
      const meanReversionStrength = 0.001;
      const randomComponent = (Math.random() - 0.5) * 2 * config.volatility;
      const meanReversion = (config.basePrice - current.price) / config.basePrice * meanReversionStrength;

      const priceChange = current.price * (randomComponent + meanReversion);
      const newPrice = Math.max(current.price + priceChange, 0.01);

      const decimals = config.decimals || 2;
      const roundedPrice = Math.round(newPrice * Math.pow(10, decimals)) / Math.pow(10, decimals);

      const change = roundedPrice - current.open;
      const changePercent = (change / current.open) * 100;

      this.currentPrices[symbol] = {
        ...current,
        price: roundedPrice,
        high: Math.max(current.high, roundedPrice),
        low: Math.min(current.low, roundedPrice),
        change: Math.round(change * 100) / 100,
        changePercent: Math.round(changePercent * 100) / 100,
        volume: current.volume + Math.floor(Math.random() * 1000),
        lastUpdate: new Date().toISOString()
      };

      // Store history (keep last 500 points)
      this.priceHistory[symbol].push({ time: Date.now(), price: roundedPrice });
      if (this.priceHistory[symbol].length > 500) {
        this.priceHistory[symbol].shift();
      }

      updates.push(this.currentPrices[symbol]);
    });

    // Emit updates
    updates.forEach(update => {
      this.emit('priceUpdate', update);
    });
  }

  getQuote(symbol) {
    return this.currentPrices[symbol] || null;
  }

  getSnapshot() {
    return { ...this.currentPrices };
  }

  getHistory(symbol, points = 100) {
    const history = this.priceHistory[symbol] || [];
    return history.slice(-points);
  }

  getAllByType(type) {
    return Object.entries(this.currentPrices)
      .filter(([_, data]) => data.type === type)
      .map(([symbol, data]) => data);
  }

  // Apply shock to prices (used when macro event releases)
  applyShock(shockConfig) {
    const { symbol, direction, magnitude } = shockConfig;

    if (!this.currentPrices[symbol]) return;

    const current = this.currentPrices[symbol];
    const shockMultiplier = direction === 'up' ? (1 + magnitude) : (1 - magnitude);
    const newPrice = current.price * shockMultiplier;

    const decimals = this.assets[symbol]?.decimals || 2;
    const roundedPrice = Math.round(newPrice * Math.pow(10, decimals)) / Math.pow(10, decimals);

    this.currentPrices[symbol] = {
      ...current,
      price: roundedPrice,
      high: Math.max(current.high, roundedPrice),
      low: Math.min(current.low, roundedPrice),
      change: roundedPrice - current.open,
      changePercent: ((roundedPrice - current.open) / current.open) * 100,
      lastUpdate: new Date().toISOString()
    };

    this.emit('priceUpdate', this.currentPrices[symbol]);
  }
}

File: services/market_data.py
"""
Market Data Service - Live prices from real APIs

Fetches real-time prices from:
- Alpha Vantage / Yahoo (stocks, ETFs - with fallback)
- CoinGecko (crypto - free, no key needed)

Optimized for fast initial load with parallel requests.
"""

import time
import threading
import requests
from datetime import datetime
from typing import Dict, List, Callable, Optional
from concurrent.futures import ThreadPoolExecutor, as_completed

# Asset configuration with realistic base prices for fallback
ASSETS = {
    # Equities
    'SPY': {'name': 'S&P 500 ETF', 'type': 'equity', 'base_price': 596.50},
    'QQQ': {'name': 'Nasdaq 100 ETF', 'type': 'equity', 'base_price': 525.80},
    'IWM': {'name': 'Russell 2000 ETF', 'type': 'equity', 'base_price': 225.40},
    'DIA': {'name': 'Dow Jones ETF', 'type': 'equity', 'base_price': 437.20},
    
    # Forex
    'EUR/USD': {'name': 'Euro/Dollar', 'type': 'fx', 'base_price': 1.0285},
    'GBP/USD': {'name': 'Pound/Dollar', 'type': 'fx', 'base_price': 1.2180},
    'USD/JPY': {'name': 'Dollar/Yen', 'type': 'fx', 'base_price': 156.50},
    'DXY': {'name': 'Dollar Index', 'type': 'fx', 'base_price': 109.35},
    
    # Bonds
    'TLT': {'name': '20+ Year Treasury ETF', 'type': 'bond', 'base_price': 87.45},
    'IEF': {'name': '7-10 Year Treasury ETF', 'type': 'bond', 'base_price': 91.20},
    'HYG': {'name': 'High Yield Bond ETF', 'type': 'bond', 'base_price': 78.65},
    
    # Volatility
    'VIX': {'name': 'CBOE Volatility Index', 'type': 'volatility', 'base_price': 15.80},
    'VVIX': {'name': 'VIX of VIX', 'type': 'volatility', 'base_price': 92.50},
    
    # Commodities
    'GLD': {'name': 'Gold ETF', 'type': 'commodity', 'base_price': 266.80},
    'USO': {'name': 'Oil Fund', 'type': 'commodity', 'base_price': 74.20},
    
    # Crypto (CoinGecko IDs)
    'BTC': {'name': 'Bitcoin', 'type': 'crypto', 'coingecko': 'bitcoin', 'base_price': 105000},
    'ETH': {'name': 'Ethereum', 'type': 'crypto', 'coingecko': 'ethereum', 'base_price': 3300},
    'SOL': {'name': 'Solana', 'type': 'crypto', 'coingecko': 'solana', 'base_price': 260},
    'XRP': {'name': 'Ripple', 'type': 'crypto', 'coingecko': 'ripple', 'base_price': 3.15},
}


class MarketDataService:
    """Fetches live market data from real APIs with fast initialization."""
    
    def __init__(self):
        self.assets = ASSETS
        self.prices: Dict[str, dict] = {}
        self.price_history: Dict[str, List[dict]] = {}
        self._running = False
        self._thread: Optional[threading.Thread] = None
        self._callbacks: List[Callable] = []
        self._last_update = None
        self._initialized = False
        
        # Initialize with base prices immediately (so UI shows something right away)
        now = datetime.utcnow()
        for symbol, config in self.assets.items():
            base_price = config.get('base_price', 100)
            self.prices[symbol] = {
                'symbol': symbol,
                'name': config['name'],
                'type': config['type'],
                'price': base_price,
                'change': 0,
                'change_percent': 0,
                'last_update': now.isoformat()
            }
            # Add initial history point
            self.price_history[symbol] = [{
                'time': int(now.timestamp() * 1000),
                'price': base_price
            }]
    
    def on_price_update(self, callback: Callable):
        """Register a callback for price updates."""
        self._callbacks.append(callback)
    
    def start_simulation(self):
        """Start fetching live prices - shows base prices immediately, then fetches real data."""
        if self._running:
            return
        
        self._running = True
        
        # Fetch live crypto prices in background (these are reliable)
        threading.Thread(target=self._fetch_crypto_immediately, daemon=True).start()
        
        # Start the periodic update loop
        self._thread = threading.Thread(target=self._price_loop, daemon=True)
        self._thread.start()
        print("Market data service started")
    
    def _fetch_crypto_immediately(self):
        """Fetch crypto prices immediately since CoinGecko is reliable and fast."""
        try:
            crypto_prices = self._fetch_coingecko_prices(['bitcoin', 'ethereum', 'solana', 'ripple'])
            if crypto_prices:
                self._update_crypto_prices(crypto_prices)
                print("Live crypto prices loaded!")
        except Exception as e:
            print(f"Initial crypto fetch error: {e}")
    
    def stop(self):
        """Stop the price fetcher."""
        self._running = False
        if self._thread:
            self._thread.join(timeout=2)
    
    def _price_loop(self):
        """Main loop that fetches prices and simulates small movements."""
        import random
        
        while self._running:
            try:
                # Fetch live crypto prices
                crypto_prices = self._fetch_coingecko_prices(['bitcoin', 'ethereum', 'solana', 'ripple'])
                if crypto_prices:
                    self._update_crypto_prices(crypto_prices)
                
                # Simulate small realistic movements for non-crypto assets
                # (since Yahoo is rate limiting, we simulate based on realistic volatility)
                self._simulate_price_movements()
                
            except Exception as e:
                print(f"Price loop error: {e}")
            
            # Update every 30 seconds
            time.sleep(30)
    
    def _update_crypto_prices(self, crypto_prices: Dict):
        """Update crypto prices from CoinGecko data."""
        now = datetime.utcnow()
        
        crypto_mapping = {
            'bitcoin': 'BTC',
            'ethereum': 'ETH', 
            'solana': 'SOL',
            'ripple': 'XRP'
        }
        
        for cg_id, symbol in crypto_mapping.items():
            if cg_id in crypto_prices:
                price = crypto_prices[cg_id]['price']
                change_pct = crypto_prices[cg_id].get('change_percent', 0)
                
                if price > 0:
                    self.prices[symbol] = {
                        'symbol': symbol,
                        'name': self.assets[symbol]['name'],
                        'type': 'crypto',
                        'price': price,
                        'change': 0,
                        'change_percent': change_pct,
                        'last_update': now.isoformat()
                    }
                    
                    self.price_history[symbol].append({
                        'time': int(now.timestamp() * 1000),
                        'price': price
                    })
                    
                    if len(self.price_history[symbol]) > 200:
                        self.price_history[symbol] = self.price_history[symbol][-200:]
    
    def _simulate_price_movements(self):
        """Simulate small realistic price movements for non-crypto assets."""
        import random
        now = datetime.utcnow()
        
        volatility = {
            'equity': 0.0015,
            'fx': 0.0003,
            'bond': 0.0008,
            'volatility': 0.02,
            'commodity': 0.001
        }
        
        for symbol, config in self.assets.items():
            if config['type'] == 'crypto':
                continue  # Crypto uses live prices
            
            vol = volatility.get(config['type'], 0.001)
            current_price = self.prices[symbol]['price']
            
            # Random walk with mean reversion
            change = random.gauss(0, vol)
            new_price = current_price * (1 + change)
            
            # Mean reversion toward base price
            base = config.get('base_price', current_price)
            reversion = (base - new_price) * 0.001
            new_price += reversion
            
            # Update price
            change_pct = ((new_price - current_price) / current_price) * 100
            
            self.prices[symbol] = {
                'symbol': symbol,
                'name': config['name'],
                'type': config['type'],
                'price': round(new_price, 4 if config['type'] == 'fx' else 2),
                'change': new_price - current_price,
                'change_percent': round(change_pct, 2),
                'last_update': now.isoformat()
            }
            
            self.price_history[symbol].append({
                'time': int(now.timestamp() * 1000),
                'price': new_price
            })
            
            if len(self.price_history[symbol]) > 200:
                self.price_history[symbol] = self.price_history[symbol][-200:]
        
        self._last_update = now
    
    def _fetch_coingecko_prices(self, coin_ids: List[str]) -> Dict[str, dict]:
        """Fetch crypto prices from CoinGecko (free, no key, reliable)."""
        prices = {}
        
        if not coin_ids:
            return prices
        
        try:
            ids_str = ','.join(coin_ids)
            url = f"https://api.coingecko.com/api/v3/simple/price?ids={ids_str}&vs_currencies=usd&include_24hr_change=true"
            
            response = requests.get(url, timeout=5)
            
            if response.status_code == 200:
                data = response.json()
                
                for coin_id in coin_ids:
                    if coin_id in data:
                        coin_data = data[coin_id]
                        prices[coin_id] = {
                            'price': float(coin_data.get('usd', 0)),
                            'change_percent': float(coin_data.get('usd_24h_change', 0))
                        }
        
        except Exception as e:
            print(f"CoinGecko error: {e}")
        
        return prices
    
    def get_quote(self, symbol: str) -> Optional[dict]:
        """Get current quote for a symbol."""
        return self.prices.get(symbol)
    
    def get_snapshot(self) -> Dict[str, dict]:
        """Get snapshot of all current prices."""
        return self.prices.copy()
    
    def get_history(self, symbol: str, points: int = 100) -> List[dict]:
        """Get price history for a symbol."""
        history = self.price_history.get(symbol, [])
        return history[-points:] if history else []
    
    def apply_shock(self, shock_config: dict):
        """Apply a price shock (for simulating event impacts)."""
        symbol = shock_config.get('symbol')
        magnitude = shock_config.get('magnitude', 0)
        
        if symbol in self.prices and self.prices[symbol]['price'] > 0:
            old_price = self.prices[symbol]['price']
            new_price = old_price * (1 + magnitude / 100)
            
            self.prices[symbol]['price'] = new_price
            self.prices[symbol]['change'] = new_price - old_price
            self.prices[symbol]['change_percent'] += magnitude
